controllers/ProposalController.php

```diff
@@ -1144,4 +1167,125 @@ private function saveEditedReview(\app\models\Review $review, $reviewStatus)
         return $review;
     }
 
+    /**
+     * Check if user can publish Proposal
+     *
+     * @param Proposal $proposal
+     * @return bool
+     */
+    private function canPublishProposal(Proposal $proposal)
+    {
+        $currentUser = MainController::getCurrentUser();
+
+        if (($currentUser->role === User::USER_ROLE_PUBLISHER || $currentUser->role === User::USER_ROLE_ADMIN)
+            && ($proposal->submitter_id !== $currentUser->id) && $proposal->status === \app\models\Proposal::STATUS_PENDING) {
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Pass pending proposal's status to rejected
+     *
+     * @return string
+     * @throws CannotSaveException
+     */
+    public function actionRejectProposal()
+    {
+        $proposalId = Yii::$app->request->post()['proposalId'];
+        $selectedProposal = $this->checkIfProposalExists($proposalId);
+        $selectedProposal->status = \app\models\Proposal::STATUS_REJECTED;
+
+        if (!$selectedProposal->save()) {
+            throw new CannotSaveException($selectedProposal);
+        }
+
+        return $this->actionProposal($proposalId);
+    }
+
+    /**
+     * Display the publishing form of a proposal
+     * If post loaded, it tweets the proposal and
+     * save proposal's status to published
+     *
+     * @param int $id
+     * @return string
+     * @throws CannotSaveException
+     * @throws FileDoesNotExistException
+     * @throws \app\models\exceptions\CannotAddMediaToTweetException
+     * @throws \app\models\exceptions\FileException
+     * @throws \app\models\exceptions\TwitterAPIException
+     * @throws \app\models\exceptions\TwitterAPIInvalidFileContentException
+     */
+    public function actionPublishProposal(int $id)
+    {
+        $selectedProposal = $this->checkIfProposalExists($id);
+        $publishProposalFormModel = new PublishProposalForm();
+
+        if ($publishProposalFormModel->load(Yii::$app->request->post()) && $publishProposalFormModel->validate()) {
+            $twitterConnector = new TwitterConnector();
+
+            if ($publishProposalFormModel->file) {
+                $twitterConnector->addMedia('../uploaded-files/proposal-related-files/'. $selectedProposal->file->path);
+            }
+
+            $twitterConnector->postTweet($publishProposalFormModel->content);
+
+            $selectedProposal->status = \app\models\Proposal::STATUS_PUBLISHED;
+
+            if (!$selectedProposal->save())
+            {
+                throw new CannotSaveException($selectedProposal);
+            }
```

Comment:
I would've done something a little different:
- Start a DB transaction
- Open a try block from line 1228 to line 1240
- Set the `published` status BEFORE tweeting (in order to avoid case when a tweet is posted and a database save exception comes just after - We cannot cancel a tweet once it is posted to the Twitter API)
- The `postTweet` function returns an array containing the HTTP status code and the response of the Twitter API. If the statusCode is not 2xx, throw an exception
- Catch any exception, rollback and re-throw catched exception or return view with an error message

Reply:
Yep good thought !